python/triton/compiler/code_generator.py

```diff
@@ -440,22 +438,24 @@ def visit_FunctionDef(self, node):
         self.builder.set_insertion_point_to_start(entry)
         # visit function body
         self.visit_compound_statement(node.body)
+
         # finalize function
+        assert not self.builder.get_insertion_block().has_terminator()
         if self.ret_type is None or self.ret_type == language.void:
             self.ret_type = language.void
             self.builder.ret([])
         else:
-            # update return type
-            if isinstance(self.ret_type, tuple):
-                self.prototype.ret_types = list(self.ret_type)
-                self.fn.reset_type(self.prototype.to_ir(self.builder))
-            else:
-                self.prototype.ret_types = [self.ret_type]
-                self.fn.reset_type(self.prototype.to_ir(self.builder))
+            self.prototype.ret_types = list(self.ret_type) if isinstance(self.ret_type, tuple) else [self.ret_type]
+            self.fn.reset_type(self.prototype.to_ir(self.builder))
+            self.builder.ret([
```

Comment:
Why there has to be a `return`?

Reply:
Isn't it already done in `visit_Return`?

Reply:
This return will terminate the `post_ret_block` created by `visit_Return`. So the IR for:
```python
def identity(x):
    return x
```
will look like:
```mlir
^bb0:
tt.return %x

^bb1:  // post_ret_block
tt.return %undef
```